streams/shared/src/main/scala/zio/stream/experimental/ZTransducer.scala

```diff
@@ -62,22 +65,55 @@ object ZTransducer {
   def chunkN[I](n: Int): ZTransducer[Any, Nothing, I, I] =
     ZTransducer {
       for {
-        buffered <- Ref.make[Chunk[I]](Chunk.empty).toManaged_
-        done     <- Ref.make(false).toManaged_
+        buffered <- ZRef.makeManaged[Chunk[I]](Chunk.empty)
         push = { (input: Option[Chunk[I]]) =>
           input match {
-            case None => done.set(true) *> buffered.getAndSet(Chunk.empty)
+            case None =>
+              buffered.modify(buf => (if (buf.isEmpty) Push.done else Push.emit(buf)) -> Chunk.empty).flatten
             case Some(is) =>
-              buffered.modify { buffered =>
-                val concat = buffered ++ is
-                if (concat.size >= n) (concat.take(n), concat.drop(n))
-                else (concat, Chunk.empty)
-              }
+              buffered.modify { buf0 =>
+                val buf = buf0 ++ is
+                if (buf.length >= n) {
+                  val (out, buf1) = buf.splitAt(n)
+                  Push.emit(out) -> buf1
+                } else
+                  Push.next -> buf
+              }.flatten
           }
         }
       } yield push
     }
 
+  def collectAllWhile[I](p: I => Boolean): ZTransducer[Any, Nothing, I, List[I]] =
+    ZTransducer {
+      for {
+        buffered <- ZRef.makeManaged[Chunk[I]](Chunk.empty)
+        push = { (in: Option[Chunk[I]]) =>
+          buffered.modify { buf0 =>
+            val buf = in.foldLeft(buf0)(_ ++ _)
+            if (buf.isEmpty)
```

Comment:
This doesn’t look right - shouldn’t we check if `in` was originally `None`?

Reply:
if it is then `buf == buf0` and if it isn't empty then my guess is we still have to go through the remainder. Is that correct?

Reply:
Yes, that sounds right. Except that if it is `None`, we want to flush the remainder anyway (even if we haven't seen an `I` that satisfies `p`).